### ISSUE TITLE
[Inductor] Float division inside tl.load in codegen results in TypeError('unexpected type fp32')

### ISSUE THREAD

**Issue by jataylo** [pytorch/pytorch#154996]
### 🐛 Describe the bug

This is a strange issue only observed locally on AMD/ROCm runs, when torch.compiling a simple dynamic shapes conv2d+batchnorm2d example. Without dynamic shapes this passes fine. Looking at the codegen, it generates (1/3)*x0 in the tl.load, resulting in a float output which fails the the tl.load. Replacing this with integer divison the triton code at least compiles and executes.

Reproducer:
```
import torch
import torch.nn as nn

class SimpleBlock(nn.Module):
    def __init__(self, width_in, width_out=224, stride=2):
        super().__init__()
        self.conv = nn.Conv2d(width_in, width_out, kernel_size=1, stride=stride, bias=False)
        self.bn = nn.BatchNorm2d(width_out)

    def forward(self, x):
        x = self.conv(x)
        x = self.bn(x)
        return x

test_cases = [
    ((2, 32, 470, 725), 32),
    ((2, 224, 235, 363), 224),
]

for input_shape, width_in in test_cases:
    model = SimpleBlock(width_in).train().to(dtype=torch.bfloat16, device='cuda')
    input_tensor = torch.randn(*input_shape, dtype=torch.bfloat16, device='cuda')
    model = torch.compile(model)
    output = model(input_tensor)
```

Stack trace:
https://gist.github.com/jataylo/aa1d39dafd4c5909d8978a485c3fa586

(note both platforms show - W0603 09:45:03.856000 879407 torch/utils/_sympy/interp.py:176] [0/1] failed while executing pow_by_natural([VR[1, int_oo], VR[-1, -1]]))

ROCm relevant codegen: (uses float div and fails)
```
tmp0 = tl.load(in_ptr0 + (((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((((r0_2 + (1/3)*x0 + (1/3)*x0*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + (1/3)*x0*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + (1/3)*x0*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2))) // (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) % (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))))) + x1*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) > (1))) + 224*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) > (1)))*((((r0_2 + (1/3)*x0 + (1/3)*x0*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + (1/3)*x0*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + (1/3)*x0*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2))) // (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + (triton_helpers.div_floor_integer((-1) + ks0,  2)) + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) % 2)) + ((r0_2 % (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))))), xmask & r0_mask, eviction_policy='evict_last', other=0.0).to(tl.float32)`
```

NV codegen: (no float div and passes)
```
tmp0 = r0_2 + x0*(triton_helpers.div_floor_integer(4 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)),  3))
tmp1 = 2 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2))
tmp2 = tmp0 < tmp1
tmp3 = tl.load(in_ptr0 + (((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((((r0_2 + x0*(triton_helpers.div_floor_integer(4 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)),  3))) // (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) % (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))))) + x1*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) > (1))) + 224*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) > (1)))*((((r0_2 + x0*(triton_helpers.div_floor_integer(4 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)),  3))) // (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + (triton_helpers.div_floor_integer((-1) + ks0,  2)) + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) % 2)) + (((r0_2 + x0*(triton_helpers.div_floor_integer(4 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)),  3))) % (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))))), xmask & r0_mask & tmp2, eviction_policy='evict_last', other=0.0).to(tl.float32)
tmp4 = tmp3.to(tl.float32)
```

What is confusing about this is the post fusion fxgraph (after_post_grad_graph_28.txt) for both platforms are identical, my hunch is that there is a divergence somewhere which causes a different evaluation in the sympy library and the AMD run exposes some numerical bug there which floats into this erroneous codegen. Attaching the tlparse output for both platforms below.

ROCm - [dedicated_log_torch_trace_13n19vyk.log](https://github.com/user-attachments/files/20566668/dedicated_log_torch_trace_13n19vyk.log)

NV - [dedicated_log_torch_trace_uxdf5qdn.log](https://github.com/user-attachments/files/20566673/dedicated_log_torch_trace_uxdf5qdn.log)

I would appreciate any pointers on how to debug this, not really sure how to work this one out. Maybe it's just a case of more sanity checks to ensure integer division inside the tl.load statements, or the symbolic representation may just be off here entirely on the AMD run.

### Versions

Versions of relevant libraries:
[pip3] mypy==1.14.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.22.4
[pip3] onnx==1.17.0
[pip3] onnxscript==0.2.2
[pip3] optree==0.13.0
[pip3] torch==2.8.0a0+git9d6f0d5
[pip3] torchvision==0.22.0a0+d02b184
[pip3] triton==3.3.0

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @aakhundov

**Comment by jataylo** [comment]
Just found https://github.com/pytorch/pytorch/issues/153375 who experienced a similar issue on  H100. I would lean towards saying this is a more fundamental codegen bug than just ROCm specific.

**Comment by eellison** [comment]
The difference in codegen is just because rocm doesn't opt in to channels last conversion. If you do `torch._inductor.config.layout_optimization = False` you get the same line :

`tmp3 = tl.load(in_ptr0 + (((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((((r0_2 + x0*(triton_helpers.div_floor_integer(4 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)),  3))) // (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) % (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))))) + x1*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) > (1))) + 224*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))) > (1)))*((1) * ((1) >= (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) + (1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) * ((1 + (triton_helpers.div_floor_integer((-1) + ks1,  2))) > (1)))*((((r0_2 + x0*(triton_helpers.div_floor_integer(4 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)),  3))) // (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + (triton_helpers.div_floor_integer((-1) + ks0,  2)) + (triton_helpers.div_floor_integer((-1) + ks1,  2)))) % 2)) + (((r0_2 + x0*(triton_helpers.div_floor_integer(4 + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks1,  2)) + 2*(triton_helpers.div_floor_integer((-1) + ks0,  2))*(triton_helpers.div_floor_integer((-1) + ks1,  2)),  3))) % (1 + (triton_helpers.div_floor_integer((-1) + ks0,  2)))))), xmask & r0_mask & tmp2, eviction_policy='evict_last', other=0.0).to(tl.float32)`

But that doesn't fail for me on h100.. 

What exactly is the expr in tl.load that gives fp32 ?

also dynamic shapes, cc @bobrenjc93 .

**Comment by jataylo** [comment]
> What exactly is the expr in tl.load that gives fp32 ?

@eellison the float division makes this return to float i.e. "(1/3) * x0" , there shouldn't be a float divison in the tl.load line.

If I find and replace 1/3 with 1 // 3 then the kernel runs.


Somehow NV in this case doesn't generate that divison but in a related issue it does https://github.com/pytorch/pytorch/issues/153375
